tests/test_categorical.py

```diff
@@ -0,0 +1,275 @@
+from math import frexp
+
+import numpy as np
+import pandas as pd
+import pytest
+from numpy.testing import assert_almost_equal
+
+from pywhy_stats import categorical
+
+from .testdata import testdata
+
+seed = 12345
+
+df_adult = pd.read_csv("tests/testdata/adult.csv")
+
+
+def _binary_scm(n_samples=200):
+    # We construct a SCM where X1 -> Y <- X and Y -> Z
+    # so X1 is independent from X, but conditionally dependent
+    # given Y or Z
+    rng = np.random.default_rng(seed)
+
+    X = rng.binomial(1, 0.3, (n_samples, 1))
+    X1 = rng.binomial(1, 0.6, (n_samples, 1))
+    Y = X * X1
+    Z = Y + (1 - rng.binomial(1, 0.5, (n_samples, 1)))
+
+    # create input for the CI test
+    df = pd.DataFrame(np.hstack((X, X1, Y, Z)), columns=["x", "x1", "y", "z"])
+    return df
+
+
+def test_chisquare_marginal_independence_adult_dataset():
+    """Test that chi-square tests return the correct answer for marginal independence queries.
+
+    Uses the test data from dagitty.
+    """
+    # Comparision values taken from dagitty (DAGitty)
+    lambda_ = "pearson"
+    X = df_adult["Age"]
+    Y = df_adult["Immigrant"]
+    result = categorical.ind(X=X, Y=Y, lambda_=lambda_)
+    assert_almost_equal(result.statistic, 57.75, decimal=1)
+    assert_almost_equal(np.log(result.pvalue), -25.47, decimal=1)
+    assert result.additional_information["dof"] == 4
+
+    Y = df_adult["Race"]
+    result = categorical.ind(X=X, Y=Y, lambda_=lambda_)
+    assert_almost_equal(result.statistic, 56.25, decimal=1)
+    assert_almost_equal(np.log(result.pvalue), -24.75, decimal=1)
+    assert result.additional_information["dof"] == 4
+
+    Y = df_adult["Sex"]
+    result = categorical.ind(X=X, Y=Y, lambda_=lambda_)
+    assert_almost_equal(result.statistic, 289.62, decimal=1)
+    assert_almost_equal(np.log(result.pvalue), -139.82, decimal=1)
+    assert result.additional_information["dof"] == 4
+
+    X = df_adult["Immigrant"]
+    Y = df_adult["Sex"]
+    result = categorical.ind(X=X, Y=Y, lambda_=lambda_)
+    assert_almost_equal(result.statistic, 0.2724, decimal=1)
+    assert_almost_equal(np.log(result.pvalue), -0.50, decimal=1)
+    assert result.additional_information["dof"] == 1
```

Comment:
Wondering if we shouldn't only use artificial data here instead? I don't see the real benefit of having real-world data as compared to artificial one in the context of a unit test where we need to know the ground truth.

Reply:
This is to ensure feature parity wrt daggity and pcalg in R. Since those have been relatively vetted, it's a good regression test. I also have artificial data here.